client/lib/i18n-utils/utils.js

```diff
@@ -110,18 +110,177 @@ export function getLocaleFromPath( path ) {
  * @returns {string} original path with new locale slug
  */
 export function addLocaleToPath( path, locale ) {
-	const urlParts = parse( path );
+	const urlParts = url.parse( path );
 	const queryString = urlParts.search || '';
 
 	return removeLocaleFromPath( urlParts.pathname ) + `/${ locale }` + queryString;
 }
 
-export function addLocaleToWpcomUrl( url, locale ) {
+const localesToSubdomains = {
+	'pt-br': 'br',
+	br: 'bre',
+	zh: 'zh-cn',
+	'zh-hk': 'zh-tw',
+	'zh-sg': 'zh-cn',
+	kr: 'ko',
+};
+
+const magnificentNonEnLocales = [
+	'es',
+	'pt-br',
+	'de',
+	'fr',
+	'he',
+	'ja',
+	'it',
+	'nl',
+	'ru',
+	'tr',
+	'id',
+	'zh-cn',
+	'zh-tw',
+	'ko',
+	'ar',
+	'sv',
+];
+
+const localesWithBlog = [ 'en', 'ja', 'es', 'pt', 'fr', 'pt-br' ];
+const localesForJetpackCom = [
+	'en',
+	'ar',
+	'de',
+	'es',
+	'fr',
+	'he',
+	'id',
+	'it',
+	'ja',
+	'ko',
+	'nl',
+	'pt-br',
+	'ro',
+	'ru',
+	'sv',
+	'tr',
+	'zh-cn',
+	'zh-tw',
+];
+const localesWithTranslatedSupport = [ 'ja', 'de', 'ru', 'pt-br', 'es' ];
+const localesWithPrivacyPolicy = [ 'en', 'fr', 'de' ];
+const localesWithCookiePolicy = [ 'en', 'fr', 'de' ];
+
+const urlLocalizationMapping = {
+	'wordpress.com': ( urlParts, localeSlug ) => {
+		if ( magnificentNonEnLocales.includes( localeSlug ) ) {
+			if ( localeSlug in localesToSubdomains ) {
+				urlParts.host = localesToSubdomains[ localeSlug ] + '.wordpress.com';
+			} else {
+				urlParts.host = localeSlug + '.wordpress.com';
+			}
+		}
+		return urlParts;
+	},
```

Comment:
I know this PR is still WIP so ignore this if you think it's a premature optimization, but if you reckon these function blocks won't change much, maybe we could do some sort of neat abstraction. 

```
const setUrlHost = ( hostname, validLocales = [] ) => ( urlParts, localeSlug ) => {
    if ( validLocales.includes( localeSlug ) ) {
        urlParts.host = `${ localesToSubdomains[ localeSlug ] || localeSlug }.${ hostname }`;
    }
    return urlParts;
};

const urlLocalizationMapping = {
	'wordpress.com': setUrlHost( magnificentNonEnLocales, 'wordpress.com' ),
    'wordpress.com/tos': setUrlHost( localesForJetpackCom, 'jetpack.com' ),
    ...

```

Reply:
This is exactly why I already asked for some eyes! Thank you, this is very useful and clever :)